### PR TITLE
chore(gatsby): Convert jobsV2 reducer to TS

### PR DESCRIPTION
**Pull request by gabrieluizramos**
## Description
In order to continue the conversion, this PR converts `reducers/jobsv2` to TS.

## Related Issues

Related to #21995

### CODE REVIEW

File: packages/gatsby/src/bootstrap/remove-stale-jobs.ts [hunk 1]
```diff
@@ -23,13 +23,18 @@ export const removeStaleJobs = (
...
   // If any of our pending jobs do not have an existing inputPath or the inputPath changed
   // we remove it from the queue as they would fail anyway
-  state.jobsV2.incomplete.forEach(({ job, plugin }: IGatsbyJobV2): void => {
-    if (isJobStale(job)) {
-      actions.push(internalActions.removeStaleJob(job.contentDigest))
-    } else {
-      actions.push(publicActions.createJobV2(job, plugin))
+  state.jobsV2.incomplete.forEach(
+    ({
+      job = { contentDigest: ``, inputPaths: [] },
```
Comment (ascorbic) [inline packages/gatsby/src/bootstrap/remove-stale-jobs.ts:28]: It's not necessary to have a default value here. There will never be undefined jobs, because the map uses job.contentDigest as the key. If the job was undefined then there would be no key available to insert it. If the job is not found in the Map then there's no need to remove or create it, so you can just return early

File: packages/gatsby/src/redux/reducers/jobsv2.ts [hunk 2]
```diff
@@ -1,10 +1,20 @@
...
+  state: IGatsbyState["jobsV2"] = {
+    incomplete: new Map(),
+    complete: new Map(),
+  },
+  action: ActionsUnion
+): IGatsbyState["jobsV2"] => {
   switch (action.type) {
     case `CREATE_JOB_V2`: {
-      const { job, plugin } = action.payload
+      const { job = jobDefaultValue, plugin } = action.payload
```
Comment (ascorbic) [inline packages/gatsby/src/redux/reducers/jobsv2.ts:17]: We don't want to create empty jobs. Better to noop than add an empty job to the queue.

File: packages/gatsby/src/redux/reducers/jobsv2.ts [hunk 3]
```diff
@@ -16,7 +26,9 @@ module.exports = (
 
     case `END_JOB_V2`: {
       const { jobContentDigest, result } = action.payload
-      const { job } = state.incomplete.get(jobContentDigest)
+      const { job = jobDefaultValue } = state.incomplete.get(
```
Comment (ascorbic) [inline packages/gatsby/src/redux/reducers/jobsv2.ts:29]: Likewise here. If the job is undefined then there's nothing to remove. You don't want to add a non-existent job to the "complete" queue.

File: packages/gatsby/src/redux/reducers/jobsv2.ts (hunk 2 above)
Comment (gabrieluizramos) [reply packages/gatsby/src/redux/reducers/jobsv2.ts:17]: So, just to confirm: here in the reducer should I return the default state if job is undefined:
```js
if (!job) return state
```

As same as the `remove-stale-job`?

File: packages/gatsby/src/bootstrap/remove-stale-jobs.ts [hunk 4]
```diff
@@ -24,6 +24,8 @@ export const removeStaleJobs = (
   // If any of our pending jobs do not have an existing inputPath or the inputPath changed
   // we remove it from the queue as they would fail anyway
   state.jobsV2.incomplete.forEach(({ job, plugin }: IGatsbyJobV2): void => {
+    if (!job) return
```
Comment (wardpeet) [inline packages/gatsby/src/bootstrap/remove-stale-jobs.ts:27]: this should never happen, the array can't hold null values

File: packages/gatsby/src/redux/reducers/jobsv2.ts [hunk 5]
```diff
@@ -1,11 +1,18 @@
...
+    incomplete: new Map(),
+    complete: new Map(),
+  },
+  action: ActionsUnion
+): IGatsbyState["jobsV2"] => {
   switch (action.type) {
     case `CREATE_JOB_V2`: {
       const { job, plugin } = action.payload
 
+      if (!job) return state
```
Comment (wardpeet) [inline packages/gatsby/src/redux/reducers/jobsv2.ts:14]: Can't be null

File: packages/gatsby/src/redux/reducers/jobsv2.ts [hunk 6]
```diff
@@ -16,7 +23,9 @@ module.exports = (
 
     case `END_JOB_V2`: {
       const { jobContentDigest, result } = action.payload
-      const { job } = state.incomplete.get(jobContentDigest)
+      const { job } = state.incomplete.get(jobContentDigest) as IGatsbyJobV2
+
+      if (!job) return state
```
Comment (wardpeet) [inline packages/gatsby/src/redux/reducers/jobsv2.ts:28]: can't be null

File: packages/gatsby/src/redux/types.ts [hunk 7]
```diff
@@ -101,9 +101,11 @@ export interface IGatsbyJobContent {
...
 
 export interface IGatsbyJobV2 {
-  job: IGatsbyJobContent
-  plugin: IGatsbyPlugin
+  job?: IGatsbyJobContent
+  plugin?: IGatsbyPlugin
   traceId?: string
+  result?: string[]
+  inputPaths?: string[]
 }
```
Comment (wardpeet) [inline packages/gatsby/src/redux/types.ts:109]: What's the reason for these changes?

File: packages/gatsby/src/redux/reducers/jobsv2.ts (hunk 6 above)
Comment (wardpeet) [inline packages/gatsby/src/redux/reducers/jobsv2.ts:28]: Could we go for this syntax instead :)
```suggestion
      if (!job) {
        return state
      }
```

Comment (blainekasten) [inline packages/gatsby/src/redux/reducers/jobsv2.ts:28]: We should actually throw an error here that says if we encountered this error it's a bug internal to Gatsby and to report an issue.

Comment (gabrieluizramos) [reply packages/gatsby/src/redux/reducers/jobsv2.ts]: Done!

Comment (gabrieluizramos) [reply packages/gatsby/src/redux/reducers/jobsv2.ts]: @blainekasten @wardpeet Just to summarize the discussion. Should we get rid of this check, since the job would not be null never or should we trow an error? What you guys think is the best approach?

File: packages/gatsby/src/redux/types.ts (hunk 7 above)
Comment (gabrieluizramos) [reply packages/gatsby/src/redux/types.ts:109]: I think that were some types that I didn't get right at the beggining, but now things are becoming more clear to me.

I'm removing the optional for the `job`, `plugin`, `result` and `inputPaths` keys inside this interface now. Since we actually set it inside the reducer we actually need the `result` and `inputPaths` in this interface.  Am I right?

File: packages/gatsby/src/redux/reducers/jobsv2.ts (hunk 6 above)
Comment (wardpeet) [reply packages/gatsby/src/redux/reducers/jobsv2.ts]: throw an error :) Technically it could never happen but you a "Map" could return undefined so we should catch it

Comment (gabrieluizramos) [reply packages/gatsby/src/redux/reducers/jobsv2.ts]: I've just pushed some changes that I think that makes sense for both comments, fixing some types and statements inside the reducer. What you guys think about it?

Comment (gabrieluizramos) [reply packages/gatsby/src/redux/reducers/jobsv2.ts]: Great! I'll update it right now.

**Comment by wardpeet** [comment]
Sorry for not being super helpful before. Long day 😅.

You should make 2 types, 1 is the jobInput & the jobResult.
https://github.com/gatsbyjs/gatsby/blob/649b4ddd55b0f618639fe7bdc5d7a6cbcc61b468/packages/gatsby/src/utils/jobs-manager.js#L64

JobResult should be just an object, it could be anything but has to be an object.

**Comment by gabrieluizramos** [comment]
@wardpeet No worries! You guys are being super helpful actually!

But I think I didn't get it right yet...
The job input looks a little bit like the `IGatsbyJobContent`, are they the same?

Thought about updating it with a BaseJobInterface. The related types would be something like:

```ts
export interface IGastbyBaseJobInterface {
  name: string
  outputDir: string
  args: Record<string, any>
}

export type IGatsbyJobContent = IGastbyBaseJobInterface & {
  inputPaths: string[]
  contentDigest: string
  plugin: {
    name: string
    version: string
    resolve: string
  }
}

export type IGatsbyJobResult = Promise<Record<string, any>>

export interface IGatsbyJobV2 {
  job: IGatsbyJobContent
  plugin: IGatsbyPlugin
  traceId?: string
  result: string[]
  inputPaths: string[]
}

```

File: packages/gatsby/src/redux/reducers/jobsv2.ts (hunk 6 above)
Comment (gabrieluizramos) [reply packages/gatsby/src/redux/reducers/jobsv2.ts]: Done! ✅

File: packages/gatsby/src/redux/reducers/jobsv2.ts [hunk 8]
```diff
@@ -1,30 +1,47 @@
...
+        throw new Error(
+          `If you encounter this error, it's probably a Gatsby internal bug. Please open an issue reporting us this.`
+        )
+      }
+
       state.incomplete.set(job.contentDigest, {
         job,
         plugin,
-      })
+      } as IGatsbyJobV2)
```
Comment (wardpeet) [inline packages/gatsby/src/redux/reducers/jobsv2.ts:23]:
```suggestion
      } as IGatsbyIncompleteJobV2 )
```

File: packages/gatsby/src/redux/reducers/jobsv2.ts [hunk 9]
```diff
@@ -1,30 +1,47 @@
...
+      }
 
       state.incomplete.delete(job.contentDigest)
 
       // inputPaths is used to make sure the job is not stale
       state.complete.set(job.contentDigest, {
         result,
         inputPaths: job.inputPaths,
-      })
+      } as IGatsbyJobV2)
```
Comment (wardpeet) [inline packages/gatsby/src/redux/reducers/jobsv2.ts:44]:
```suggestion
      } as IGatsbyCompleteJobV2)
```

File: packages/gatsby/src/redux/types.ts [hunk 10]
```diff
@@ -249,6 +251,32 @@ export type ActionsUnion =
...
+  | ICreateJobV2Action
+  | IEndJobV2Action
+  | IRemoveStaleJobV2Action
+
+export interface ICreateJobV2Action {
+  type: `CREATE_JOB_V2`
+  payload: {
+    job: IGatsbyJobV2["job"]
+    plugin: IGatsbyJobV2["plugin"]
+  }
```
Comment (wardpeet) [inline packages/gatsby/src/redux/types.ts:263]:
```suggestion
  payload: IGatsbyCreateJobV2
```

File: packages/gatsby/src/redux/types.ts [hunk 11]
```diff
@@ -249,6 +251,32 @@ export type ActionsUnion =
...
+}
+
+export interface IEndJobV2Action {
+  type: `END_JOB_V2`
+  payload: {
+    jobContentDigest: string
+    result: string[]
```
Comment (wardpeet) [inline packages/gatsby/src/redux/types.ts:270]:
```suggestion
    result: JobResultInterface
```

**Comment by gabrieluizramos** [comment]
@wardpeet Awesome! Thank you very much! I've already pushed the changes.

File: packages/gatsby/src/redux/reducers/jobsv2.ts [hunk 12]
```diff
@@ -11,22 +16,30 @@ export const jobsV2Reducer = (
     case `CREATE_JOB_V2`: {
       const { job, plugin } = action.payload
 
-      if (!job) return state
+      if (!job) {
+        throw new Error(
+          `If you encounter this error, it's probably a Gatsby internal bug. Please open an issue reporting us this.`
+        )
+      }
```
Comment (wardpeet) [inline packages/gatsby/src/redux/reducers/jobsv2.ts:23]: This doesn't need a check, typings says it's not nullable so that's 👍 We'll have to handle this cause inside the redux/actions/public.js
```suggestion

```

Comment (gabrieluizramos) [reply packages/gatsby/src/redux/reducers/jobsv2.ts]: Nice! I've committed the changes directly from your suggestion.

**Comment by gabrieluizramos** [comment]
@wardpeet Thank you very much! You guys are awesome!!!

**Comment by gabrieluizramos** [comment]
Hi @ascorbic ! I think that maybe you still need to approve this PR. Can you help us?
